### PR TITLE
Add testing agent to CI; snapshot redis integration

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
This patch adds a testing agent which will allow us to do proper integration tests for our `contrib` integrations.

## Snapshot testing
One of the primary testing features built into the testing agent aside from testing all the trace invariants imaginable is snapshotting. Similar to how snapshot testing works [in the web development world](https://jestjs.io/docs/en/snapshot-testing), it allows us to capture the state of the generated traces of our integrations which _should not change_ unless we intentionally mean for it to. We are essentially doing snapshot testing right now with the tests for our integrations but it is tedious and cumbersome.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -126,7 +126,8 @@ def get_exts_for(name):
         install_requires=[
             "enum34; python_version<'3.4'",
             "funcsigs>=1.0.0; python_version=='2.7'",
-            "msgpack>=0.5.0",
+            "msgpack>=0.5.0,<1.0; python_version<'3.0'",
```
Comment (Kyle-Verhoog) [inline setup.py:129]: I had to force this as it turns out that the tracer when run with Python 2 under `tox` with `msgpack==1.0` causes some bizarre encoding behaviour. This strange encoding actually makes the traces undecodable by both the test agent and the real agent.

File: tests/contrib/redis/test.py [hunk 2]
```diff
@@ -193,81 +119,48 @@ def test_patch_unpatch(self):
...
-        assert dd_span.span_type == "redis"
-        assert dd_span.error == 0
-        assert dd_span.get_metric("out.redis_db") == 0
-        assert dd_span.get_tag("out.host") == "localhost"
-        assert dd_span.get_tag("redis.raw_command") == u"GET cheese"
-        assert dd_span.get_metric("redis.args_length") == 2
-        assert dd_span.resource == "GET cheese"
-
     @BaseTracerTestCase.run_in_subprocess(env_overrides=dict(DD_SERVICE="mysvc"))
+    @snapshot()
```
Comment (Kyle-Verhoog) [inline tests/contrib/redis/test.py:132]: The snapshot here has to come before the `run_in_subprocess` which we should probably try to add a check for if possible in `@snapshot()`.

File: .circleci/config.yml [hunk 3]
```diff
@@ -14,6 +14,15 @@ python35_image: &python35_image circleci/python:3.5
 python27_image: &python27_image circleci/python:2.7
 ddtrace_dev_image: &ddtrace_dev_image datadog/dd-trace-py:latest
 datadog_agent_image: &datadog_agent_image datadog/docker-dd-agent:latest
+test_agent: &test_agent
+  - image: kyleverhoog/dd-trace-test-agent:latest
```
Comment (Kyle-Verhoog) [inline .circleci/config.yml:18]: note to self: pin this in case for consistent builds

Comment (jd) [reply .circleci/config.yml:18]: and please use a Dockerfile from this repo or a repo where it's clear we maintain/can modify.

File: setup.py (hunk 1 above)
Comment (jd) [reply setup.py:129]: Add a test in a different PR with this?

File: .circleci/config.yml [hunk 4]
```diff
@@ -14,6 +14,15 @@ python35_image: &python35_image circleci/python:3.5
...
+    environment:
+      - SNAPSHOT_DIR=dd-trace-py/tests/snapshots
+      - SNAPSHOT_CI=1
```
Comment (brettlangdon) [inline .circleci/config.yml:21]: comment here would be nice

File: .circleci/config.yml [hunk 5]
```diff
@@ -14,6 +14,15 @@ python35_image: &python35_image circleci/python:3.5
...
+    # CircleCI does not provide a way to mount a volume between containers or a
+    # way to bake the snapshots into a custom image...
+    # So we have to clone the repo and snapshots into the container.
+    command: /bin/bash -c "git clone https://github.com/datadog/dd-trace-py/; cd dd-trace-py; git checkout << pipeline.git.branch >>; cd ..; java -jar app-standalone.jar"
```
Comment (brettlangdon) [inline .circleci/config.yml:25]: hmmm....

would we be better off using a different executor where we might be able to do this?

File: setup.py (hunk 1 above)
Comment (brettlangdon) [reply setup.py:129]: weird... should we try to get #1491 merged in, since that will make this issue moot?

that draft pr should mostly be good to go, no?

File: tests/contrib/redis/test.py (hunk 2 above)
Comment (brettlangdon) [reply tests/contrib/redis/test.py:132]: `@run_in_subprocess(snapshot=True)` ?

not sure if it'll be too obscure of an option or not

File: tests/contrib/redis/test.py [hunk 6]
```diff
@@ -193,81 +119,48 @@ def test_patch_unpatch(self):
...
 
         self.reset()
 
         # Manual override
         Pin.override(self.r, service="mysvc", tracer=self.tracer)
         self.r.get("cheese")
-        span = self.get_spans()[0]
-        assert span.service == "mysvc", span.service
 
+    @snapshot()
```
Comment (brettlangdon) [inline tests/contrib/redis/test.py:157]: ordering here backwards?

File: tests/util.py [hunk 7]
```diff
@@ -18,3 +23,48 @@ def override_global_tracer(tracer):
...
+
+            module = inspect.getmodule(f)
+            token = "{}{}{}.{}".format(module.__name__, "." if clsname else "", clsname, f.__name__)
+
+            try:
+                conn = httplib.HTTPConnection(tracer.writer.api.hostname, tracer.writer.api.port)
+                conn.request("GET", "/test/start?token=%s" % token)
+                r = conn.getresponse()
+                if r.status != 200:
+                    raise SnapshotFailed("", r.read().decode())
```
Comment (brettlangdon) [inline tests/util.py:53]: Error message?

File: tests/util.py [hunk 8]
```diff
@@ -18,3 +23,48 @@ def override_global_tracer(tracer):
...
+                self = args[0]
+                clsname = self.__class__.__name__
+            else:
+                clsname = ""
+
+            module = inspect.getmodule(f)
+            token = "{}{}{}.{}".format(module.__name__, "." if clsname else "", clsname, f.__name__)
+
+            try:
+                conn = httplib.HTTPConnection(tracer.writer.api.hostname, tracer.writer.api.port)
```
Comment (brettlangdon) [inline tests/util.py:49]: comments in here might be nice, like... what is the purpose of this first API call? of the second?, why do we have to force the queue to flush?

File: tests/util.py [hunk 9]
```diff
@@ -18,3 +23,48 @@ def override_global_tracer(tracer):
...
+
+                ret = f(*args, **kwargs)
+                tracer.writer.flush_queue()
+
+                ignoresqs = ",".join(ignores)
+                conn = httplib.HTTPConnection(tracer.writer.api.hostname, tracer.writer.api.port)
+                conn.request("GET", "/test/snapshot?ignores=%s&token=%s" % (ignoresqs, token))
+                r = conn.getresponse()
+                if r.status != 200:
+                    raise SnapshotFailed("", r.read().decode())
```
Comment (brettlangdon) [inline tests/util.py:63]: same here, would a message be nice?

File: .circleci/config.yml (hunk 5 above)
Comment (Kyle-Verhoog) [reply .circleci/config.yml:25]: Not quite sure what you mean 🤔. Can you clarify?

File: tests/contrib/redis/test.py (hunk 6 above)
Comment (Kyle-Verhoog) [reply tests/contrib/redis/test.py:157]: good catch!

File: tests/util.py (hunk 7 above)
Comment (Kyle-Verhoog) [reply tests/util.py:53]: The error message is passed along as an argument to the exception. Although I'm not sure why I do that versus just using the message argument itself 🤔. I'll check that again

File: .circleci/config.yml (hunk 5 above)
Comment (brettlangdon) [reply .circleci/config.yml:25]: this is gross, can we do something else?

Comment (Kyle-Verhoog) [reply .circleci/config.yml:25]: lol agreed. But see the commit history for my attempts at doing something else 😛. I'm totally open to anything other than this.

File: .circleci/config.yml [hunk 10]
```diff
@@ -17,7 +17,9 @@ datadog_agent_image: &datadog_agent_image datadog/docker-dd-agent:latest
 test_agent: &test_agent
   - image: kyleverhoog/dd-trace-test-agent:latest
     environment:
+      # Use the cloned snapshot directory (see below)
       - SNAPSHOT_DIR=dd-trace-py/tests/snapshots
+      # Enable CI mode for the test agent
```
Comment (brettlangdon) [inline .circleci/config.yml:22]: well, what is `CI` mode? 😆

Comment (Kyle-Verhoog) [reply .circleci/config.yml:22]: Maybe just worth pointing to test agent documentation when it's available. CI mode is probably going to evolve and not worth duping the information here which would become outdated

File: .circleci/config.yml (hunk 3 above)
Comment (Kyle-Verhoog) [reply .circleci/config.yml:18]: :+1: definitely! This will be the case once the repo's public 🙂

File: setup.py (hunk 1 above)
Comment (Kyle-Verhoog) [reply setup.py:129]: Yeah, made the change just here to be able to test the CI properly. I'll look into whether this is reproducible or not and open a subsequent PR. Also #1491 will solve this.

File: tests/contrib/redis/test.py (hunk 2 above)
Comment (Kyle-Verhoog) [reply tests/contrib/redis/test.py:132]: Hmm yeah, not sure about that one but it might save some headaches down the line for people copy-pasting test cases which would make it worth it IMO.

**Review by majorgreys — CHANGES_REQUESTED** [review]
thanks for keeping the current redis tests while we try out the new snapshot test! one last requested change :)

File: tests/contrib/redis/test_redis.py [hunk 11]
```diff
@@ -270,3 +271,145 @@ def test_service_precedence(self):
...
+    @snapshot()
+    def test_long_command(self):
+        self.r.mget(*range(1000))
+
+    @snapshot()
+    def test_basics(self):
+        us = self.r.get("cheese")
+        assert us is None
+
+    def test_analytics_without_rate(self):
```
[the remark this answers is not in the document]
Comment (Kyle-Verhoog) [reply tests/contrib/redis/test_redis.py:300]: I feel like maayyybe there was a reason, but don't remember 😆

**Review by majorgreys — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

File: scripts/run-tox-scenario [hunk 12]
```diff
@@ -5,4 +5,4 @@ shift
 
 # CircleCI has a bug in its workspace code where it can't handle filenames with some chars
 CLEANED_PATTERN=`echo $PATTERN | tr '^?()$' '_'`
-exec tox -l | grep "$PATTERN" | tr '\n' ',' | xargs -I ARGS tox --result-json /tmp/"$CLEANED_PATTERN".results -e ARGS -- $@
+exec tox -l | grep "$PATTERN" | tr '\n' ',' | xargs -I ARGS tox -e ARGS -- $@
```
Comment (brettlangdon) [inline scripts/run-tox-scenario:8]: we should do this in a different PR, unless it is related here?

Comment (Kyle-Verhoog) [reply scripts/run-tox-scenario:8]: sure, i think it was supposed to be anyway but it never happened upstream

**Review by majorgreys — APPROVED** [review]
